hand-eye-calibration/full_hand_eye_sample.py

```diff
@@ -0,0 +1,486 @@
+    """
+
+    settings = zivid.Settings()
+    settings.iris = 20
+    settings.exposure_time = datetime.timedelta(microseconds=10000)
+    settings.gain = 1
+    settings.brightness = 1.8
+    settings.filters.gaussian.enabled = 1
+    cam.settings = settings
+
+
+def _read_robot_state(con: rtde):
+    """Recieve robot output recipe
+
+    Args:
+        con: Connection between computer and robot
+
+    Returns:
+        image_number: Image number
+        robot_pose: Pose of robot
+    """
+
+    robot_state = con.receive()
+
+    image_number = robot_state.output_int_register_24
+    robot_pose = robot_state.output_int_register_24
+
+    return image_number, robot_pose
+
+
+def pose_from_datastring(datastring: str):
+    """Extract pose from q_r -yaml file saved by openCV
+
+    Args:
+        datastring: String of text from .yaml file
+
+    Returns:
+        pose_matrix: Robotic pose as q_r zivid Pose class
+    """
+
+    string = datastring.split("data:")[-1].strip().strip("[").strip("]")
+    pose_matrix = np.fromstring(string, dtype=np.float, count=16, sep=",").reshape(
+        (4, 4)
+    )
+    return zivid.handeye.Pose(pose_matrix)
+
+
+def _save_hand_eye_results(save_dir: Path, transform: np.array, residuals: list):
+    """Save transformation and residuals to folder
+
+    Args:
+        save_dir: Path to where data will be saved
+        transform: 4x4 transformation matrix
+        residuals: List of residuals
+    """
+
+    file_storage_transform = cv2.FileStorage(
+        str(save_dir / f"transformation.yaml"), cv2.FILE_STORAGE_WRITE
+    )
+    file_storage_transform.write("PoseState", transform)
+    file_storage_transform.release()
+
+    file_storage_residuals = cv2.FileStorage(
+        str(save_dir / f"residuals.yaml"), cv2.FILE_STORAGE_WRITE
+    )
+    residual_list = []
+    for res in residuals:
+        tmp = list([res.translation, res.translation])
+        residual_list.append(tmp)
+
+    file_storage_residuals.write(
+        "Per pose residuals for rotation in deg and translation in mm",
+        np.array(residual_list),
+    )
+    file_storage_residuals.release()
+
+
+def _generate_dataset(con: rtde, input_data):
+    """ Generate dataset based on predefined robot poses
+
+    Args:
+        con: Connection between computer and robot
+        input_data: Input package containing the specific input data registers
+
+    returns:
+        save_dir: Location where dataset is saved
+
+    Universal Robot registers:
+        output_int_register_24 = image_count:
+            Counter from robot side of number of images and poses taken.
+        output_bit_register_64 = start_capture:
+            Bool that trigger camera to capture an image
+
+        Signals to UR robot:
+        input_bit_register_64 = finish_capture:
+            Bool that trigger UR robot to move to next position.
+        input_bit_register_65 = ready_to_capture:
+            Bool that tells UR robot that the camera is ready to be used.
```

Comment:
You could consider defining this in code, perhaps as a dictionary. Would make the code where you use it more readable, as opposed to just having magic numbers.

Something like:
```
ur_registers = 
{
    "start_capture": "output_int_register_24",
    "finish_capture": "input_bit_register_64",
    ....
}

if getattr(robot_state, ur_registers["start_capture"]):
    ....
```
and `settattr()` to write.

Reply:
Done, created helper function for each instead.